package-res/resources/web/pentaho/visual/base/model.js

```diff
@@ -45,37 +73,287 @@ define([
+         *
+         * @fires "will:select"
+         * @fires "did:select"
+         * @fires "rejected:select"
+         *
+         * @see #_runPipeline
+         * @see #_changeSelection
+         * @see #_broadcastResult
+         * @see pentaho.visual.base.types.selectionModes
+         */
+        selectAction: function(proposedDataFilter, keyArgs) {
+          var selectionMode = O.getOwn(keyArgs, "selectionMode");
+          var dataFilter = proposedDataFilter;
+
+          var result = this._runPipeline(WillSelect, RejectedSelect, dataFilter, selectionMode);
+          if(result) {
+            if(result.error) return result;
+            selectionMode = result.value.selectionMode;
+            dataFilter = result.value.dataFilter;
+          }
+
+          try {
+            result = this._changeSelection(dataFilter, selectionMode);
+          } catch(e) {
+            result = new ActionResult(undefined, e);
+          }
+          return this._broadcastResult(DidSelect, RejectedSelect, result, dataFilter);
+        },
+
+        /**
+         * Updates the current selection based on a filter and on a selection mode.
+         *
+         * This method is invoked by [selectAction]{@link pentaho.visual.base.Model#selectAction} once all
+         * listeners of the ["will:select"]{@link pentaho.visual.base.events.WillSelect} event have eventually
+         * mutated the original input filter and the selection mode to be used.
+         *
+         * @param {!pentaho.data.filter.AbstractFilter} proposedDataFilter -
+         * @param {!function} proposedSelectionMode -
+         * @return {pentaho.lang.ActionResult}
+         * If unsuccessful, the `error` field contains the exception that originated the error.
+         * If successful,  the `error` field is `null` and the `value` field contains the updated current selection.
+         *
+         * @fires "will:change:select"
+         * @fires "did:change:select"
+         * @fires "rejected:change:select"
+         *
+         * @see #selectAction
+         * @see #_runPipeline
+         * @see #_broadcastResult
+         * @protected
+         */
+        _changeSelection: function(proposedDataFilter, proposedSelectionMode) {
+          var selectionMode = proposedSelectionMode || this.getv("selectionMode");
+          var dataFilter = proposedDataFilter;
+
+          var result = this._runPipeline(WillChangeSelection, RejectedChangeSelection, dataFilter, selectionMode);
+          if(result) {
+            if(result.error) return result;
+            selectionMode = result.value.selectionMode;
+            dataFilter = result.value.dataFilter;
           }
-        })
-        .implement({meta: bundle.structured});
+
+          try {
+            var currentFilter = this.get("selectionFilter");
+            var combinedSelection = selectionMode(currentFilter, dataFilter);
```

Comment:
`combinedSelection` - the use of Selection without Filter for referring to a Filter should be banned. How many times have we talked about this?